src/aranges.rs

```diff
@@ -0,0 +1,249 @@
+#![deny(missing_docs)]
+
+use endianity::{Endianity, EndianBuf};
+use parser::{parse_u16, parse_uN_as_u64, parse_unit_length, parse_debug_info_offset, parse_address_size, DebugInfoOffset, Format, ParseResult, Error};
+use std::cmp::{Ordering};
+use std::marker::PhantomData;
+use std::rc::Rc;
+
+/// The `DebugAranges` struct represents the DWARF address range information
+/// found in the `.debug_aranges` section.
+#[derive(Debug, Clone, Copy)]
+pub struct DebugAranges<'input, Endian>
+    where Endian: Endianity
+{
+    debug_aranges_section: EndianBuf<'input, Endian>,
+}
+
+impl<'input, Endian> DebugAranges<'input, Endian>
+    where Endian: Endianity
+{
+    /// Construct a new `DebugAranges` instance from the data in the `.debug_aranges`
+    /// section.
+    ///
+    /// It is the caller's responsibility to read the `.debug_aranges` section and
+    /// present it as a `&[u8]` slice. That means using some ELF loader on
+    /// Linux, a Mach-O loader on OSX, etc.
+    ///
+    /// ```
+    /// use gimli::{DebugAranges, LittleEndian};
+    ///
+    /// # let buf = [];
+    /// # let read_debug_aranges_section_somehow = || &buf;
+    /// let debug_aranges = DebugAranges::<LittleEndian>::new(read_debug_aranges_section_somehow());
+    /// ```
+    pub fn new(debug_aranges_section: &'input [u8]) -> DebugAranges<'input, Endian> {
+        DebugAranges { debug_aranges_section: EndianBuf(debug_aranges_section, PhantomData) }
+    }
+
+    /// Iterate the aranges in the `.debug_aranges` section.
+    /// ```
+    /// use gimli{DebugAranges, LittleEndian};
+    ///
+    /// # let buf = [];
+    /// # let read_debug_aranges_section_somehow = || &buf;
+    /// let debug_aranges = DebugAranges::<LittleEndian>::new(read_debug_aranges_section_somehow());
+    ///
+    /// for parse_result in debug_aranges.aranges() {
+    ///   let arange = parse_result.unwrap();
+    ///   println!("arange starts at {}, has length {}", arange.start(), arange.len());
+    /// }
+    /// ```
+    pub fn aranges(&self) -> ArangesIter<Endian> {
+        ArangesIter { input: self.debug_aranges_section }
+    }
+}
+
+pub struct ArangesIter<'input, Endian>
+    where Endian: Endianity
+{
+    input: EndianBuf<'input, Endian>
+}
+
+impl<'input, Endian> Iterator for ArangesIter<'input, Endian>
+    where Endian: Endianity
+{
+    type Item = ParseResult<Vec<Arange>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.input.is_empty() {
+            None
+        } else {
+            match Arange::parse(self.input) {
+                Ok((remaining_input, aranges)) => {
+                    self.input = remaining_input;
+                    Some(Ok(aranges))
+                }
+                Err(e) => {
+                    self.input = self.input.range_to(..0);
+                    Some(Err(e))
```

Comment:
Fallible iterators are kind of hard to get right (see https://github.com/fitzgen/gimli/issues/21) so we've been switching from iterators that return `Option<ParseResult<T>>` to `next_whatever` methods that return `ParseResult<Option<T>>` which subtly encourage robustness in the face of errors. Basically, it is easy to keep iterating without handling an error, when as soon as we yeild an error we can't meaningfully yield anything else after that.